### PR TITLE
Update Cargo and its dependencies

### PR DESCRIPTION
**Pull request by alexcrichton**
This'll probably have a bunch of build errors, so let's try and head those off
and find them sooner rather than later!

### CODE REVIEW

**Comment by EdSchouten** [comment]
Thanks for working on this! Including `cc 1.0.4` is going to be pretty useful for me to get automated builds for CloudABI to work.

**Comment by kennytm** [comment]
```
[01:26:02] cargo/guide/index.html:103: broken link - cargo/guide/build-cache.html
[01:26:02] cargo/print.html:187: broken link - cargo/guide/build-cache.html
```

**Comment by alexcrichton** [comment]
I think that should be fixed w/ https://github.com/rust-lang/cargo/pull/4926 (and hopefully good to go afterwards)

**Comment by kennytm** [comment]
Travis is green, so r=me after fixing the merge conflict (and let the actual debugging begin 😄).

**Comment by kennytm** [comment]
As expected

```
  = note: C:\projects\rust\build\x86_64-pc-windows-gnu\stage2-tools\x86_64-pc-windows-gnu\release\deps\liblibgit2_sys-8baa38816dcee9e8.rlib(socket_stream.c.obj):socket_stream.c:(.text$socket_connect+0x187): undefined reference to `gai_strerrorA'
```

**Comment by alexcrichton** [comment]
Ok looks like this is because libgit2 is compiled with MinGW's gcc compiler which presumably has a definition for `gai_strerror` that depends on the symbol `gai_strerrorA`. The linker, however, is picking up the library from the [`winapi-x86_64-pc-windows-gnu` package](https://crates.io/crates/winapi-x86_64-pc-windows-gnu) which does not have this symbol.

@retep998 do you know where [the x86_64 version of `libws2_32.a` came from](https://github.com/retep998/winapi-rs/blob/0.3/x86_64/lib/libws2_32.a)? Is it perhaps from a different version of a MinGW compiler thatn we're using?

**Comment by retep998** [comment]
The import libraries that I ship in `winapi-x86_64-pc-windows-gnu` are import libraries that I made myself. They export the same set of symbols that the import libraries in the Windows SDK do. I don't use the ones from MinGW-w64 because they don't have any symbols from newer versions of Windows.

Looking at `gai_strerrorA`, for `pc-windows-msvc` it is defined as an *inline* function in the headers and it is not exported by any import libraries. MinGW-w64 meanwhile has that function as part of `libws2_32.a` so that library is in fact a combination of both static symbols *and* dynamic symbols.

I'm not entirely sure what the best solution here is. I don't want to use MinGW-w64 import libraries because of how outdated they are, but I also recognize that they have a lot of symbols which are necessary when linking C code compiled with MinGW. One option is to provide a way to opt out of the bundled import libraries when you want to use the ones from an external MinGW toolchain.

**Comment by retep998** [comment]
Another possible solution is for me to prefix all the libraries that winapi bundles so instead of `libws2_32.a` I would provide `libwin_ws2_32.a` (prefix still to be decided), so that way I could get functions from newer versions of Windows, without conflicting with import libraries provided by rustc or an external MinGW toolchain.

**Comment by alexcrichton** [comment]
Ok thanks for the info, since these are winapi specific (I think?) Thee namesapacing seems like the way to go. I may override locally in this repo temporarily while we wait on that if you're ok with that course of action

**Comment by alexcrichton** [comment]
re-r? @kennytm

**Comment by kennytm** [comment]
Wait no, Travis failed. 

@bors r-

File: src/tools/winapi-patch/i686/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1 @@
+// See the `README.md` in the directory above this for why this crate exists
```
Comment (kennytm) [inline src/tools/winapi-patch/i686/src/lib.rs]:
```
[00:03:42] tidy error: /checkout/src/tools/winapi-patch/x86_64/src/lib.rs: incorrect license
[00:03:42] tidy error: /checkout/src/tools/winapi-patch/i686/src/lib.rs: incorrect license
[00:03:43] some tidy checks failed
```

File: src/Cargo.toml [hunk 2]
```diff
@@ -55,8 +55,23 @@ debug-assertions = false
...
 
 [patch.crates-io]
+# Similar to Cargo above we want the RLS to use a vendored version of `rustfmt`
+# that we're shipping as well (to ensure that the rustfmt in RLS and the
+# `rustfmt` executable are the same exact vesion). Unlike Cargo, however, the
+# RLS depends on `rustfmt` from crates.io, so we put this in a `[patch]` section
+# for crates.io
 rustfmt-nightly = { path = "tools/rustfmt" }
+
+# Temporary measures to fix comiple on MinGW. See `tools/winapi-patch/README.md`
```
Comment (kennytm) [inline src/Cargo.toml]: Please also fix the typo as well 😸

**Comment by kennytm** [comment]
r=me after tidy passed.

**Comment by kennytm** [comment]
Can’t build `mdbook` this time. https://ci.appveyor.com/project/rust-lang/rust/build/1.0.6043/job/vh807oj55f72qit1. Why are these two libraries needed o_O

```
[01:01:10]   = note: ld: cannot find -ldbghelp
[01:01:10]           ld: cannot find -lmsimg32
[01:01:10]           
[01:01:10] 
[01:01:10] error: aborting due to previous error
[01:01:10] 
[01:01:10] error: Could not compile `mdbook`.
```

- - -

The test actually failed after 1 hour, but the status is not reported back to bors, causing it to wait 10 hours (hard-coded in homu). This is the second time this happened this month. We should reduce the bors wait time to ~3 hours.

**Comment by retep998** [comment]
Oh boy, it failed *earlier* this time. Does the version of MinGW-w64 you're using *not* come with `libdbghelp.a` or `libmsimg32.a`? You're using `backtrace` which explains the former dependency and *something* is using stuff from the `wingdi` module in `winapi` which would result in the latter dependency. 

I'll try to get out a new version today with prefixed import libraries to hopefully resolve this.

**Comment by alexcrichton** [comment]
I think that was https://github.com/rust-lang/rust/issues/47359 where we're not copying libdbghelp.a into the sysroot. That should be updated now.

@bors: r=kennytm

**Comment by kennytm** [comment]
Same error.

**Comment by retep998** [comment]
It's building the things using stage0 and that stage doesn't yet have the fix of those two import libraries in the sysroot.

**Comment by kennytm** [comment]
Ah. Could we build them using stage2?

**Comment by Mark-Simulacrum** [comment]
I would prefer that we don't move anything to later stages because stage 0 doesn't work; if this is a bug in stage 0, we can backport and bump the bootstrap compiler or workaround it temporarily, I think. However, if there's no other solution, moving to stage 2 compilation *should* be relatively painless.

**Comment by alexcrichton** [comment]
I've confirmed that the master branch of winapi builds cleanly (no env vars set) with MinGW due to the use of prefixed libraries. Due to other bugs however we can't actually use `[patch]` to depend on git repositories (it fails distcheck), @retep998 would you be able to make a release soon?

**Comment by retep998** [comment]
@alexcrichton You will still need to set one of those environment variables though in order to avoid issues with external projects that depend on rustc internal crates failing to link because they can't find winapi's prefixed libraries. I've published new versions now.

**Pull request merged**